README.md
# Learning with Texts

[![Latest Stable Version](https://poser.pugx.org/hugofara/lwt/v)](https://packagist.org/packages/hugofara/lwt)
[![License](https://poser.pugx.org/hugofara/lwt/license)](https://packagist.org/packages/hugofara/lwt)
[![PHP Version Require](https://poser.pugx.org/hugofara/lwt/require/php)](https://packagist.org/packages/hugofara/lwt)
![Composer Continuous Integration](https://github.com/hugofara/lwt/actions/workflows/php.yml/badge.svg)
[![Docker Image](https://github.com/HugoFara/lwt/actions/workflows/docker-image.yml/badge.svg)](https://github.com/HugoFara/lwt/actions/workflows/docker-image.yml)
[![Discord Server](https://badgen.net/discord/members/zAE8GXMKFa?icon=discord)](https://discord.gg/zAE8GXMKFa)

**Learning with Texts** (LWT) is a tool for language learning by reading. It is a self-hosted web application.

<div style="text-align: center;">
<img src="https://github.com/HugoFara/lwt/raw/master/img/lwt_icon_big.jpg" />
</div>

You feel that you won't learn much by translating dumb sentences or using grammar books? Learning With Texts offers you the possibility to learn by reading texts in your target language. Concept: when you don't know a word in a text, just click it. We show you the translation, and you will have regular tests to remember it. Ready to go?

**THIS IS A THIRD PARTY VERSION**. This version is not the
official one, and brings many improvements and new features.
It is quicker, has smaller database size,
and is open for contributions. The official version is on
[source forge](https://sourceforge.net/projects/learning-with-texts)

## Installation

As it is self-hosted, you will need a server, which can be your computer.

* Get the [latest GitHub release](https://github.com/HugoFara/lwt/releases).
  * You can also try to download the [latest stable version](https://github.com/HugoFara/lwt/archive/refs/heads/master.zip) if you want the cutting-edge updates (that may include some bugs).

### Linux

Start a shell in the downloaded folder an run: ``./INSTALL.sh``. You may need to run ``chmod +x ./INSTALL.sh`` first.

### Other Systems

* **Please see**: [docs/install.md](docs/install.md) for detailed instructions.
* Create ``connect.inc.php`` with an existing database user. Everything is explained at [docs/info.html](https://hugofara.github.io/lwt/docs/info.html#install).

And you are ready to go!

Please note that a composer repo is available, as well as an installation using Docker.

## Description

LWT is a language learning web application. To learn a language, you
need to practice, and we guide you in reading exercises.

First copy/paste any text you want to read. It can be raw text or an RSS feed.

![Adding French text](https://github.com/HugoFara/lwt/raw/master/img/05.jpg)

Then, we parse the text. Unknown words will be displayed with different colors,
just click them to see it in a dictionary.

![Learning French text](https://github.com/HugoFara/lwt/raw/master/img/06.jpg)

Read as much as you want!

To make sure you memorize new words, you can take review exercises.

![Reviewing French word](https://github.com/HugoFara/lwt/raw/master/img/07.jpg)

The difference with popular remembering software like
[Anki](https://apps.ankiweb.net/) is that we keep track of the
context to help you. By the way, we also ship
an Anki exporter.

## Features

* **Full list**: [docs/features.md](docs/features.md)

Features included from the official LWT software:

* Support for almost 40 languages.
* Text parsing for roman languages, right-to-left,
and East-Asian ideogram systems
* Translate words on-the-fly
* Add an audio track and read it online
* Practice words you don't remember
* Statistics to record your progress

### Features not in the official LWT

* **Full list**: [docs/newfeatures.md](docs/newfeatures.md)

Features that were added by the community:

* Support for mobile
* Automatically import texts from RSS feeds
* Support for different themes
* Display translations of terms with status in the reading frame
* Multiwords selection (click and hold on a word
→ move to another word → release mouse button)
* Bulk translate new words in the reading frame
* Text to speech
* Optional "ignore all" button in read texts
* Key bindings in the reading frame
* Selecting terms according to a text tag
* Two database backup modes (new or old structure)

### Improvements compared to the official LWT

* Database improvements (db size is much smaller now)
* Longer (>9) expressions can now be saved (up to 250 characters)
* Save text/audio position in the reading frame
* Google api (use 'ggl.php' instead of '*<http://translate.google.com>' for Google Translate)
* Improved Search/Query for Words/Texts
* Term import with more options (i.e.: combine translations, multiple tag import)
* Support for MeCab for Japanese word-by-word automatic translation.
* You can include video files from popular video platforms.
* Code documentation.
* Code is well organised, making debugging and contribution easier.

## Contribute

* **Complete explanation**: [docs/contribute.md](docs/contribute.md)

To contribute, you need to clone or fork this repository, and [Composer](https://getcomposer.org/download/).
The composer package is at [hugofara/lwt](https://packagist.org/packages/hugofara/lwt).

Run ``git clone https://github.com/HugoFara/lwt``

Next, got to the lwt folder and use ``composer install --dev``.

In short:

```bash
git clone https://github.com/HugoFara/lwt
cd lwt
composer install --dev
```

## Branches

* The stable branch is *master*. Last commit on this branch is
considered to be bug-free.
* The *dev* branch is for unstable versions.
* The *official* branch is for the official LWT Releases.
Any other branch if considered under development.

## Useful links

* General documentation at [docs/info.html](https://hugofara.github.io/lwt/docs/info.html).
* Please find more help at [docs/index.html](https://hugofara.github.io/lwt/docs/index.html).
* You can also contact the community using [GitHub](https://github.com/hugofara/lwt) or
[Discord](https://discord.gg/zAE8GXMKFa).

## Alternatives

* *See also*: [docs/links.md](docs/links.md)
LWT is written for web, but you have many choices if you want a standalone installation:
* [LingL](https://github.com/gustavklopp/LingL) is a wonderful alternative written in Python, by [gustavklopp](https://github.com/gustavklopp).
* [FLTR ◆ Foreign Language Text Reader](https://sourceforge.net/projects/foreign-language-text-reader/), a Java clone, by the same
author as LWT ([lang-learn-guy](https://sourceforge.net/u/lang-learn-guy/profile/))

## Unlicense

Under unlicense, view [LICENSE.md](LICENSE.md), please look at [http://unlicense.org/].

**Let's learn new languages!**


docs/CHANGELOG.md
# Changelog

This project's changelog. Versions marked with "-fork" come from the community, other versions come from the canonical LWT ("official" branch on Git).
For git tags, official releases are marked like "v1.0.0", while unofficial ones are marked like "v1.0.0-fork".

## [Unreleased]

### Added

* The translation and romanization of new words can now be automatic.
  * Supported automatic translation is achieved using [LibreTranslate](https://libretranslate.com/).
  * Romanization using [MeCab](https://taku910.github.io/mecab/) (Japanese only) toward katakana only.
* A lot of improvements for language creation/edition: 
  * More intuitive fields, better interactions with the user, etc...
  * A "Pop-Up" checkbox helps you choose if the website should be displayed in a pop-up.
* LWT becomes easier to install and maintain for users:
  * An intaller for Linux users at ``INSTALL.sh`` helps installing requirements and configuration.
  * Database creation wizard (``database_wizard.php``) to help setting the connection with the database.
  * A server data page at ``server_data.php`` showing all relevant information.
* In ``index.php``:
  * ``get_server_data_table`` replaces ``get_server_data`` as a better designed function.
  * ``index_do_main_page`` now renders the page to avoid global variables.
  * You get warnings if the PHP version is too low, or if a newer version of LWT is released.
* ``pagestart_kernel_nobody`` in ``inc/kernel.php`` that displays the minimal HTML formatting.
* ``inc/classes/`` folder for a better incorporation of OOP.
  * ``inc/classes/Term`` defines the ``Term`` class representing a word or multi-word.
  * ``inc/classes/Language`` defines the ``Language`` class representing a language.

### Changed

* Graphic review:
  * Headers reviewed. Most h4 headers became div.bigger, and h3 became h1 level headers.
  * We now use responsive design instead of fixed-size.
  * All themes updated accordingly, so themes were modified.
  * ``do_text_page`` (``do_text.php``) and ``do_test_page`` (``do_test.php``) now enlarge the body.
* ``echo_lwt_logo`` in ``inc/session_utility.php`` echoes a logo, without information on the table set used
because it was useless. This information is now displayed on the welcome page.
* Access to the database prefix edition at ``start.php`` was reviewed and simplified. It is still considered a legacy feature.
* Language creation reviewed:
  * When creating a new language, the language wizard do no longer display in a pop-up but in the main window.
  * Select option boxes replaced by checkboxes (better accessibility).
* Updated [hoverIntent jQuery Plug-in](http://briancherne.github.io/jquery-hoverIntent/)
from 1.8.0 to 1.10.2. This brings some bug fixes.
* "TO DO" changed to a more explicit "Unknown words" in text header.
* Refactored ``bulk_translate_word.php``. It has a better visual aspect and works better.
* Updated composer dependencies.
  * vimeo/psalm updated from ^4.12 to ^5.6.
  * phpunit/phpunit updated from ^9.5 to ^10.0.
* It is recommended to use proper URL for all dictionaries and translators.
  * '*' before a dictionary URL should be replaced by an argument 'lwt_popup=true'
  * 'ggl.php' and 'glosbe_api.php' should be updated (e. g. "{YOUR_LWT_CURRENT_ADDRESS}/ggl.php")
  * Marking the location of a term to translate with '###' is deprecated in favor of 'lwt_term'.
* **LICENSE**: we changed from The Unlincense (Unlicense) to The MIT License (MIT). Please view [post]() for detailed information. 

### Fixed

* Markdown: enforced consistency with official style recomendations. The documentation display got improved.
* The default date for new words was note accepted depending on the SQL configuration, causing issues with Docker installation. See [#78](https://github.com/HugoFara/lwt/issues/78).
* Bulk translate words:
  * Translating with bulk translate was not updating word rendering.
  * All broom icons icon were display in the screen top-right corner, and not
  at the right if text field.
* Main page (``index.php``) was not rendered properly on non-default theme.
* Full-text speech dispatcher was not available when the name of the language name was changed. Issue signalled [with issue #80](https://github.com/HugoFara/lwt/issues/80#issuecomment-1368180304).
* Support for PHP 7.4 restored, wrapper for ``str_starts_with``.
* The encoding of ``docs/info.html`` is finally fixed!

## 2.6.0-fork (January 01 2023)

### Added


inc/kernel_utility.php
<?php
/**
 * \file
 * \brief Core utility functions that do not require a complete session.
 * 
 * @package Lwt
 * @author  HugoFara <hugo.farajallah@protonmail.com>
 * @license Unlicense <http://unlicense.org/>
 * @link    https://hugofara.github.io/lwt/docs/html/kernel__utility_8php.html
 * @since   2.0.3-fork
 */

 require __DIR__ . '/settings.php';


/**
 * Return LWT version for humans
 *
 * Version is hardcoded in this function.
 * For instance 1.6.31 (October 03 2016)
 *
 * @global bool $debug If true adds a red "DEBUG"
 *
 * @return string Version number HTML-formatted
 *
 * @psalm-return '2.5.3-fork (November 06 2022) <span class="red">DEBUG</span>'|'2.5.3-fork (November 06 2022)'
 */
function get_version(): string 
{
    global $debug;
    $version = '2.6.0-fork (January 01 2023)'; 
    if ($debug) {
        $version .= ' <span class="red">DEBUG</span>';
    }
    return $version;
}

/**
 * Return a machine readable version number.
 *
 * @return string Machine-readable version, for instance v001.006.031 for version 1.6.31.
 */
function get_version_number(): string 
{
    $r = 'v';
    $v = get_version();
    // Escape any detail like "-fork"
    $v = preg_replace('/-\w+\d*/', '', $v);
    $pos = strpos($v, ' ', 0);
    if ($pos === false) { 
        my_die('Wrong version: '. $v); 
    }
    $vn = preg_split("/[.]/", substr($v, 0, $pos));
    if (count($vn) < 3) { 
        my_die('Wrong version: '. $v); 
    }
    for ($i=0; $i<3; $i++) { 
        $r .= substr('000' . $vn[$i], -3); 
    }
    return $r;
}

/**
 * Escape special HTML characters.
 * 
 * @param  string $s String to escape.
 * @return string htmlspecialchars($s, ENT_COMPAT, "UTF-8");
 */
function tohtml($s) 
{
    if (!isset($s)) {
        return ''; 
    }
    return htmlspecialchars($s, ENT_COMPAT, "UTF-8");
}


/**
 * Echo debugging informations.
 */
function showRequest(): void 
{
    $olderr = error_reporting(0);
    echo "<pre>** DEBUGGING **********************************\n";
    echo '$GLOBALS...'; 
    print_r($GLOBALS);
    echo 'get_version_number()...'; 
    echo get_version_number() . "\n";
    echo 'get_magic_quotes_gpc()...'; 
    echo "NOT EXISTS (FALSE)\n";
    echo "********************************** DEBUGGING **</pre>";
    error_reporting($olderr);
}

/**
 * Get the time since the last call
 * 
 * @return float Time sonce last call
 */
function get_execution_time()
{
    static $microtime_start = null;
    if ($microtime_start === null) {
        $microtime_start = microtime(true);
        return 0.0;
    }
    return microtime(true) - $microtime_start;
}

/**
 * Reload $setting_data if necessary
 * 
 * @return array $setting_data
 */
function get_setting_data() 
{
    static $setting_data;
    if (!$setting_data) {
        $setting_data = array(
        'set-text-h-frameheight-no-audio' => array(
            "dft" => '140', "num" => 1, "min" => 10, "max" => 999
        ),
        'set-text-h-frameheight-with-audio' => array(
            "dft" => '200', "num" => 1, "min" => 10, "max" => 999
        ),
        'set-text-l-framewidth-percent' => array(
            "dft" => '60', "num" => 1, "min" => 5, "max" => 95
        ),
        'set-text-r-frameheight-percent' => array(
            "dft" => '37', "num" => 1, "min" => 5, "max" => 95
        ),
        'set-test-h-frameheight' => array(
        	"dft" => '140', "num" => 1, "min" => 10, "max" => 999
        ),
        'set-test-l-framewidth-percent' => array(
        	"dft" => '50', "num" => 1, "min" => 5, "max" => 95
        ),
        'set-test-r-frameheight-percent' => array(
        	"dft" => '50', "num" => 1, "min" => 5, "max" => 95
        ),
        'set-words-to-do-buttons' => array(
        	"dft" => '1', "num" => 0
        ),
        'set-tooltip-mode' => array(
        	"dft" => '2', "num" => 0
        ),
        'set-display-text-frame-term-translation' => array(
        	"dft" => '1', "num" => 0
        ),
        'set-text-frame-annotation-position' => array(
        	"dft" => '2', "num" => 0
        ),
        'set-test-main-frame-waiting-time' => array(
        	"dft" => '0', "num" => 1, "min" => 0, "max" => 9999
        ),
        'set-test-edit-frame-waiting-time' => array(
        	"dft" => '500', "num" => 1, "min" => 0, "max" => 99999999
        ),
        'set-test-sentence-count' => array(
        	"dft" => '1', "num" => 0
        ),
        'set-tts' => array(
        	"dft" => '1', "num" => 0
        ),
        'set-term-sentence-count' => array(
        	"dft" => '1', "num" => 0
        ),
        'set-archivedtexts-per-page' => array(
        	"dft" => '100', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-texts-per-page' => array(
        	"dft" => '10', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-terms-per-page' => array(
        	"dft" => '100', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-tags-per-page' => array(
        	"dft" => '100', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-articles-per-page' => array(
        	"dft" => '10', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-feeds-per-page' => array(
        	"dft" => '50', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-max-articles-with-text' => array(
        	"dft" => '100', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-max-articles-without-text' => array(
        	"dft" => '250', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-max-texts-per-feed' => array(
        	"dft" => '20', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-ggl-translation-per-page' => array(
        	"dft" => '100', "num" => 1, "min" => 1, "max" => 9999
        ),
        'set-regex-mode' => array(
        	"dft" => '', "num" => 0
        ),
        'set-theme_dir' => array(
        	"dft" => 'themes/default/', "num" => 0
        ),
        'set-text-visit-statuses-via-key' => array(
        	"dft" => '', "num" => 0
        ),
        'set-term-translation-delimiters' => array(
        	"dft" => '/;|', "num" => 0
        ),
        'set-mobile-display-mode' => array(
        	"dft" => '0', "num" => 0
        ),
        'set-similar-terms-count' => array(
        	"dft" => '0', "num" => 1, "min" => 0, "max" => 9)
        );
    }
    return $setting_data;
}

/**
 * Remove all spaces from a string.
 * 
 * @param string $s      Input string
 * @param string $remove Do not do anything if empty
 * 
 * @return string String without spaces if requested.
 */
function remove_spaces($s, $remove) 
{
    if (!$remove) { 
        return $s;
    }
    return str_replace(' ', '', $s);  // '' enthält &#x200B;
}

/**
 * Returns path to the MeCab application.
 * MeCab can split Japanese text word by word
 *
 * @param string $mecab_args Arguments to add
 *
 * @return string OS-compatible command
 *
 * @since 2.3.1-fork Much more verifications added
 */
function get_mecab_path($mecab_args = ''): string 
{
    $os = strtoupper(substr(PHP_OS, 0, 3));
    $mecab_args = escapeshellcmd($mecab_args);
    if ($os == 'LIN') {
        if (shell_exec("command -v mecab")) {
            return 'mecab' . $mecab_args; 
        }
        my_die("MeCab not detected! Please install it and add it to your PATH.");
    }
    if ($os == 'WIN') {
        if (shell_exec('where /R "%ProgramFiles%\\MeCab\\bin" mecab.exe')) { 
            return '"%ProgramFiles%\\MeCab\\bin\\mecab.exe"' . $mecab_args;
        } 
        if (shell_exec('where /R "%ProgramFiles(x86)%\\MeCab\\bin" mecab.exe')) {
            return '"%ProgramFiles(x86)%\\MeCab\\bin\\mecab.exe"' . $mecab_args; 
        }
        if (shell_exec('where mecab.exe')) {
            return 'mecab.exe' . $mecab_args; 
        }
        my_die("MeCab not detected! Install it or add it to the PATH.");
    }
    my_die("Your OS '$os' cannot use MeCab with this version of LWT!");
}


/**
 * Find end-of-sentence characters in a sentence using latin alphabet.
 * 
 * @param string[] $matches       All the matches from a capturing regex
 * @param string   $noSentenceEnd If different from '', can declare that a string a not the end of a sentence.
 * 
 * @return string $matches[0] with ends of sentences marked with \t and \r.
 */
function find_latin_sentence_end($matches, $noSentenceEnd)
{
    if (!strlen($matches[6]) && strlen($matches[7]) && preg_match('/[a-zA-Z0-9]/', substr($matches[1], -1))) { 
        return preg_replace("/[.]/", ".\t", $matches[0]); 
    }
    if (is_numeric($matches[1])) {
        if (strlen($matches[1]) < 3) { 
            return $matches[0];
        }
    } else if ($matches[3] && (preg_match('/^[B-DF-HJ-NP-TV-XZb-df-hj-np-tv-xz][b-df-hj-np-tv-xzñ]*$/u', $matches[1]) || preg_match('/^[AEIOUY]$/', $matches[1]))
    ) { 
        return $matches[0]; 
    }
    if (preg_match('/[.:]/', $matches[2]) && preg_match('/^[a-z]/', $matches[7])) {
        return $matches[0];
    }
    if ($noSentenceEnd != '' && preg_match("/^($noSentenceEnd)$/", $matches[0])) {
        return $matches[0]; 
    }
    return $matches[0] . "\r";
}


/**
 * Make the script crash and prints an error message
 *
 * @param string $text Error text to output
 *
 * @return never
 * 
 * @since 2.5.3-fork Add a link to the Discord community
 */
function my_die($text) 
{
    echo '</select></p></div><div style="padding: 1em; color:red; font-size:120%; background-color:#CEECF5;">' .
    '<p><b>Fatal Error:</b> ' . 
    tohtml($text) . 
    "</p></div><hr /><pre>Backtrace:\n\n";
    debug_print_backtrace();
    echo '</pre><hr />
    <p>Signal this issue on 
    <a href="https://github.com/HugoFara/lwt/issues/new/choose">GitHub</a> or 
    <a href="https://discord.gg/xrkRZR2jtt">Discord</a>.</p>';
    die('</body></html>');
}

/**
 * Display the main menu of navigation as a dropdown
 */
function quickMenu(): void 
{
    ?>

<select id="quickmenu" onchange="quickMenuRedirection(value)">
    <option value="" selected="selected">[Menu]</option>
    <option value="index">Home</option>
    <optgroup label="Texts">
        <option value="edit_texts">Texts</option>
        <option value="edit_archivedtexts">Text Archive</option>
        <option value="edit_texttags">Text Tags</option>
        <option value="check_text">Text Check</option>
        <option value="long_text_import">Long Text Import</option>
    </optgroup>
    <option value="edit_languages">Languages</option>
    <optgroup label="Terms">
        <option value="edit_words">Terms</option>
        <option value="edit_tags">Term Tags</option>
        <option value="upload_words">Term Import</option>
    </optgroup>
    <option value="statistics">Statistics</option>
    <option value="rss_import">Newsfeed Import</option>
    <optgroup label="Other">
        <option value="backup_restore">Backup/Restore</option>
        <option value="settings">Settings</option>
        <option value="text_to_speech_settings">Text-to-Speech Settings</option>
        <option value="INFO">Help</option>
    </optgroup>
</select>
    <?php
}

/**
 * Start a page without connecting to the database with a complete header and a non-closed body.
 *
 * @param string $title  Title of the page
 * @param string $addcss Some CSS to be embed in a style tag
 *
 * @global string $tbpref The database table prefix if true
 * @global int    $debug  Show the requests if true
 */
function pagestart_kernel_nobody($title, $addcss=''): void 
{
    global $tbpref, $debug;
    @header('Expires: Wed, 11 Jan 1984 05:00:00 GMT');
    @header('Last-Modified: ' . gmdate('D, d M Y H:i:s') . ' GMT');
    @header('Cache-Control: no-cache, must-revalidate, max-age=0');
    @header('Pragma: no-cache');
    ?><!DOCTYPE html>
    <?php 
    echo '<html lang="en">';
    ?>
<head>
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <!-- 
        <?php echo file_get_contents("LICENSE.md");?> 
    -->
    <meta name="viewport" content="width=900" />
    <link rel="shortcut icon" href="favicon.ico" type="image/x-icon"/>
    <link rel="apple-touch-icon" href="img/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon" sizes="72x72" href="img/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon" sizes="114x114" href="img/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-startup-image" href="img/apple-touch-startup.png" />
    <meta name="apple-mobile-web-app-capable" content="yes" />
    
    <link rel="stylesheet" type="text/css" href="css/jquery-ui.css" />
    <link rel="stylesheet" type="text/css" href="css/jquery.tagit.css" />
    <link rel="stylesheet" type="text/css" href="css/styles.css" />
    <link rel="stylesheet" type="text/css" href="css/feed_wizard.css" />
    <style type="text/css">
        <?php echo $addcss . "\n"; ?>
    </style>
    
    <script type="text/javascript" src="js/jquery.js" charset="utf-8"></script>
    <script type="text/javascript" src="js/jquery.scrollTo.min.js" charset="utf-8"></script>
    <script type="text/javascript" src="js/jquery-ui.min.js"  charset="utf-8"></script>
    <script type="text/javascript" src="js/jquery.jeditable.mini.js" charset="utf-8"></script>
    <script type="text/javascript" src="js/tag-it.js" charset="utf-8"></script>
    <script type="text/javascript" src="js/overlib/overlib_mini.js" charset="utf-8"></script>
    <!-- URLBASE : "<?php echo tohtml(url_base()); ?>" -->
    <!-- TBPREF  : "<?php echo tohtml($tbpref);  ?>" -->
    <script type="text/javascript">
        //<![CDATA[
        var STATUSES = <?php echo json_encode(get_statuses()); ?>;
        //]]>
    </script>
    
    <title>LWT :: <?php echo tohtml($title); ?></title>
</head>
    <?php
    echo '<body>';
    ?>
<div id="overDiv" style="position:absolute; visibility:hidden; z-index:1000;"></div>
    <?php
    flush();
    if ($debug) { 
        showRequest(); 
    }
}

/**
 * Add a closing body tag.
 *
 * @global bool $debug Show the requests if true
 * @global float $dspltime Total execution time since the PHP session started
 */
function pageend(): void 
{
    global $debug, $dspltime;
    if ($debug) { 
        showRequest(); 
    }
    if ($dspltime) { 
        echo "\n<p class=\"smallgray2\">" . 
        round(get_execution_time(), 5) . " secs</p>\n"; 
    }
    echo '</body></html>';
} 

/**
 * Debug function only.
 *
 * @param mixed  $var  A printable variable to debug
 * @param string $text Echoed text in HTML page
 *
 * @global bool $debug This functions doesn't do anything is $debug is false.
 */
function echodebug($var,$text): void 
{
    global $debug;
    if ($debug) { 
        echo "<pre> **DEBUGGING** " . tohtml($text) . ' = [[[';
        print_r($var);
        echo "]]]\n--------------</pre>";
    }
}


/**
 * Return an associative array of all possible statuses 
 * 
 * @return array<int<1, 5>|98|99, array<string, string>> 
 * Statues, keys are 1, 2, 3, 4, 5, 98, 99. 
 * Values are associative arrays of keys abbr and name 
 */  
function get_statuses() 
{
    static $statuses;
    if (!$statuses) {
        $statuses = array(
        1 => array("abbr" =>   "1", "name" => "Learning"),
        2 => array("abbr" =>   "2", "name" => "Learning"),
        3 => array("abbr" =>   "3", "name" => "Learning"),
        4 => array("abbr" =>   "4", "name" => "Learning"),
        5 => array("abbr" =>   "5", "name" => "Learned"),
        99 => array("abbr" => "WKn", "name" => "Well Known"),
        98 => array("abbr" => "Ign", "name" => "Ignored"),
        );
    }
    return $statuses;
}

/**
 * Replace the first occurence of $needle in $haystack by $replace
 * 
 * @param  string $needle   Text to replace
 * @param  string $replace  Text to replace by
 * @param  string $haystack Input string
 * @return string String with replaced text
 */
function str_replace_first($needle, $replace, $haystack) 
{
    if ($needle === '') {
        return $haystack; 
    }
    $pos = strpos($haystack, $needle);
    if ($pos !== false) {
        return substr_replace($haystack, $replace, $pos, strlen($needle));
    }
    return $haystack;
}

/**
 * Convert annotations in a JSON format.
 * 
 * @param  string $ann Annotations.
 * @return string A JSON-encoded version of the annotations
 */
function annotation_to_json($ann) 
{
    if ($ann == '') {
        return "{}"; 
    }
    $arr = array();
    $items = preg_split('/[\n]/u', $ann);
    foreach ($items as $item) {
        $vals = preg_split('/[\t]/u', $item);
        if (count($vals) > 3 && $vals[0] >= 0 && $vals[2] > 0) {
            $arr[intval($vals[0])-1] = array($vals[1], $vals[2], $vals[3]);
        }
    }
    return json_encode($arr);
}

/**
 * Get a request when possible. Otherwise, return an empty string.
 * 
 * @param  string $s Request key
 * @return string Trimmed request or empty string
 */
function getreq($s) 
{
    if (isset($_REQUEST[$s]) ) {
        return trim($_REQUEST[$s]);
    } else {
        return ''; 
    }
}

/**
 * Get a session variable when possible. Otherwise, return an empty string.
 * 
 * @param  string $s Session variable key
 * @return string Trimmed sesseion variable or empty string
 */
function getsess($s) 
{
    if (isset($_SESSION[$s]) ) {
        return trim($_SESSION[$s]);
    } 
    return '';
}

/**
 * Get the base URL of the application
 *
 * @return string base URL
 */
function url_base(): string 
{
    $url = parse_url("http://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]");
    $r = $url["scheme"] . "://" . $url["host"];
    if (isset($url["port"])) { 
        $r .= ":" . $url["port"]; 
    }
    if(isset($url["path"])) {
        $b = basename($url["path"]);
        if (substr($b, -4) == ".php" || substr($b, -4) == ".htm" || substr($b, -5) == ".html") { 
            $r .= dirname($url["path"]); 
        }
        else {
            $r .= $url["path"]; 
        }
    }
    if (substr($r, -1) !== "/") { 
        $r .= "/"; 
    }
    return $r;
}


/**
 * Make a random score for a new word.
 *
 * @param 'iv'|'id'|'u'|string $type Type of insertion
 *                                   * 'iv': Keys only (TodayScore, Tomorrow, Random)
 *                                   * 'id': Values only
 *                                   * 'u': Key = value pairs
 *
 * @return string SQL code to use
 */
function make_score_random_insert_update($type): string 
{
    // $type='iv'/'id'/'u'
    if ($type == 'iv') {
        return ' WoTodayScore, WoTomorrowScore, WoRandom ';
    } 
    if ($type == 'id') {
        return ' ' . getsqlscoreformula(2) . ', ' . getsqlscoreformula(3) . ', RAND() ';
    } 
    if ($type == 'u') {
        return ' WoTodayScore = ' . getsqlscoreformula(2) . ', WoTomorrowScore = ' . getsqlscoreformula(3) . ', WoRandom = RAND() ';
    } 
    return '';
}

/**
 * SQL formula for computing score.
 *
 * @param int $method Score for tomorrow (2), the day after it (3) or never (any value).
 *
 * @return string SQL score computation string
 *
 * @psalm-return '
        GREATEST(-125, CASE 
            WHEN WoStatus > 5 THEN 100
            WHEN WoStatus = 1 THEN ROUND(-7 * DATEDIFF(NOW(),WoStatusChanged))
            WHEN WoStatus = 2 THEN ROUND(6.9 - 3.5 * DATEDIFF(NOW(),WoStatusChanged))
            WHEN WoStatus = 3 THEN ROUND(20 - 2.3 * DATEDIFF(NOW(),WoStatusChanged))
            WHEN WoStatus = 4 THEN ROUND(46.4 - 1.75 * DATEDIFF(NOW(),WoStatusChanged))
            WHEN WoStatus = 5 THEN ROUND(100 - 1.4 * DATEDIFF(NOW(),WoStatusChanged))
        END)'|'
        GREATEST(-125, CASE 
            WHEN WoStatus > 5 THEN 100 
            WHEN WoStatus = 1 THEN ROUND(-7 -7 * DATEDIFF(NOW(),WoStatusChanged)) 
            WHEN WoStatus = 2 THEN ROUND(3.4 - 3.5 * DATEDIFF(NOW(),WoStatusChanged)) 
            WHEN WoStatus = 3 THEN ROUND(17.7 - 2.3 * DATEDIFF(NOW(),WoStatusChanged)) 
            WHEN WoStatus = 4 THEN ROUND(44.65 - 1.75 * DATEDIFF(NOW(),WoStatusChanged)) 
            WHEN WoStatus = 5 THEN ROUND(98.6 - 1.4 * DATEDIFF(NOW(),WoStatusChanged)) 
        END)'|'0'
 */
function getsqlscoreformula($method): string 
{
    // 
    // Formula: {{{2.4^{Status}+Status-Days-1} over Status -2.4} over 0.14325248}
        
    if ($method == 3) { 
        return '
        GREATEST(-125, CASE 
            WHEN WoStatus > 5 THEN 100 
            WHEN WoStatus = 1 THEN ROUND(-7 -7 * DATEDIFF(NOW(),WoStatusChanged)) 
            WHEN WoStatus = 2 THEN ROUND(3.4 - 3.5 * DATEDIFF(NOW(),WoStatusChanged)) 
            WHEN WoStatus = 3 THEN ROUND(17.7 - 2.3 * DATEDIFF(NOW(),WoStatusChanged)) 
            WHEN WoStatus = 4 THEN ROUND(44.65 - 1.75 * DATEDIFF(NOW(),WoStatusChanged)) 
            WHEN WoStatus = 5 THEN ROUND(98.6 - 1.4 * DATEDIFF(NOW(),WoStatusChanged)) 
        END)';
    }
    if ($method == 2) { 
        return '
        GREATEST(-125, CASE 
            WHEN WoStatus > 5 THEN 100
            WHEN WoStatus = 1 THEN ROUND(-7 * DATEDIFF(NOW(),WoStatusChanged))
            WHEN WoStatus = 2 THEN ROUND(6.9 - 3.5 * DATEDIFF(NOW(),WoStatusChanged))
            WHEN WoStatus = 3 THEN ROUND(20 - 2.3 * DATEDIFF(NOW(),WoStatusChanged))
            WHEN WoStatus = 4 THEN ROUND(46.4 - 1.75 * DATEDIFF(NOW(),WoStatusChanged))
            WHEN WoStatus = 5 THEN ROUND(100 - 1.4 * DATEDIFF(NOW(),WoStatusChanged))
        END)';
    } 
    return '0';
}


/**
 * Display a error message vanishing after a few seconds.
 * 
 * @param string $msg Message to display.
 * @param bool $noback If true, don't display a button to go back
 * 
 * @return string HTML-formatted string for an automating vanishing message.
 */
function error_message_with_hide($msg, $noback): string 
{
    if (trim($msg) == '') { 
        return ''; 
    }
    if (substr($msg, 0, 5) == "Error" ) {
        return '<p class="red">*** ' . tohtml($msg) . ' ***' . 
        ($noback ? 
        '' : 
        '<br /><input type="button" value="&lt;&lt; Go back and correct &lt;&lt;" onclick="history.back();" />' ) . 
        '</p>'; 
    } 
    return '<p id="hide3" class="msgblue">+++ ' . tohtml($msg) . ' +++</p>'; 
}

/***************** Wrappers for PHP <8.0  ********************/
if (!function_exists('str_starts_with')) {
    function str_starts_with($haystack, $needle) {
        return substr($haystack, 0, strlen($needle)) == $needle;
    }
}

if (!function_exists('str_contains')) {
    function str_contains($haystack, $needle) {
        return strpos($haystack, $needle) !== false;
    }
}

?>